sdk/storage/AzureStorageBlob/Source/Credentials/StorageCredentials.swift

```diff
@@ -29,23 +29,102 @@ import Foundation
+    internal var sasToken: String? {
+        didSet { expirationDate = StorageSASCredential.expiration(forToken: sasToken) }
+    }
+
+
+    // Treat tokens that expire in <= 2 minutes as expired
+    private static let expirationOffset: TimeInterval = 2 * 60
+
+    public var expirationDate: Date?
+    public var expired: Bool {
+        guard let expirationDate = expirationDate else { return true }
+        return (expirationDate - StorageSASCredential.expirationOffset) >= Date()
+    }
+
+    // MARK: Static methods
+
+    private static func expiration(forToken sasToken: String?) -> Date? {
+        guard let token = sasToken else { return nil }
+        guard let comps = URLComponents(string: "?\(token)") else { return nil }
+        guard let expiration = comps.queryItems?.filter({ $0.name == "se" }).first?.value else { return nil }
+        return Date(expiration, format: .iso8601)
+    }
+
+    // MARK: Initializers
 
     /// Create a shared access signature credential from an account-level shared access signature.
     /// - Parameters:
-    ///   - connectionString: An account-level shared access signature connection string.
-    public init(connectionString: String) {
+    ///   - connectionStringProvider: A closure that returns an account-level shared access signature connection string.
+    public init(connectionStringProvider: @escaping () -> String) {
+        self.blobSasUriProvider = nil
+        self.connectionStringProvider = connectionStringProvider
+        clear()
+    }
+
+    /// Create a shared access signature credential from a container- or blob-level shared access signature.
+    /// - Parameters:
+    ///   - blobSasUriProvider: A closure that returns a container- or blob-level shared access signature URI.
+    public init(blobSasUriProvider: @escaping () -> String) {
+        self.blobSasUriProvider = blobSasUriProvider
+        self.connectionStringProvider = nil
+        clear()
+    }
+
+    /// Create a shared access signature credential from an account-level shared access signature.
+    /// - Parameters:
+    ///   - connectionString: The initial account-level shared access signature connection string.
+    ///   - connectionStringProvider: A closure that returns an account-level shared access signature connection string.
+    public init(connectionString: String, connectionStringProvider: (() -> String)? = nil) {
+        self.blobSasUriProvider = nil
+        self.connectionStringProvider = connectionStringProvider
+        self.error = configureFrom(connectionString: connectionString)
+    }
+
+    /// Create a shared access signature credential from a container- or blob-level shared access signature.
+    /// - Parameters:
+    ///   - blobSasUri: The initial container- or blob-level shared access signature URI.
+    ///   - blobSasUriProvider: A closure that returns a container- or blob-level shared access signature URI.
+    public init(blobSasUri: String, blobSasUriProvider: (() -> String)? = nil) {
```

Comment:
Is it possible to provide just a provider and not an initial value... where it would simply need to contact the provider initially? Same question for the connection string variety. Or is it that the initial connection string contains some important info that the connection string provider can't retrieve?

Reply:
Yep, those are above. I couldn't collapse them into a single initializer each because then both parameters would be optional and a customer could choose to provide neither of them. So now you can either provide a credential provider closure, or a static credential and an optional credential provider to be used to refresh the credential.

Reply:
I actually debated removing the version with the static credential entirely, since we know tokens WILL expire eventually, so a static credential will eventually require to be refreshed. But I figured there might be scenarios where a developer makes a token that expires far in the future because they simply don't want to ever have to deal with the complexity of refreshing it, so they treat it like a static key.

Reply:
Recommend reaching out to @adrianhall on this question.

Reply:
Discussed, we will ditch these initializers and go with only the version that takes the callback.